modules/service/src/main/resources/db/migration/V0847__observation_reference.sql

```diff
@@ -0,0 +1,92 @@
+
+    BEGIN
+        EXECUTE 'CREATE SEQUENCE IF NOT EXISTS ' || sequence_name;
+    EXCEPTION WHEN unique_violation THEN
+        NULL; -- sequence exists, this was an attempt to create it in parallel
+    END;
+
+    RETURN nextval(sequence_name)::INT;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Update all existing observations so that they have an index defined.
+UPDATE t_observation
+  SET c_observation_index = next_observation_index(c_program_id);
+
+-- Make the index required.
+ALTER TABLE t_observation
+  ALTER COLUMN c_observation_index SET NOT NULL;
+
+-- Add a trigger that sets the observation index when a new observation is
+-- created.
+CREATE OR REPLACE FUNCTION set_observation_index()
+RETURNS TRIGGER AS $$
+BEGIN
+  NEW.c_observation_index = next_observation_index(NEW.c_program_id);
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER set_observation_index_trigger
+BEFORE INSERT ON t_observation
+FOR EACH ROW
+EXECUTE FUNCTION set_observation_index();
+
+-- Prevent the observation index from changing.
+CREATE OR REPLACE FUNCTION prevent_observation_index_update()
+RETURNS TRIGGER AS $$
+BEGIN
+  IF NEW.c_observation_index != OLD.c_observation_index THEN
+    RAISE EXCEPTION 'The observation index (in column c_observation_index) cannot be updated once it is set';
+  END IF;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER prevent_observation_index_update_trigger
+BEFORE UPDATE on t_observation
+FOR EACH ROW
+EXECUTE FUNCTION prevent_observation_index_update();
+
+
+-- Update the observation v to include the new column.
+-- Re-create this view to include the new index column.
+DROP VIEW v_observation;
+CREATE VIEW v_observation AS
+  SELECT o.*,
+  CASE WHEN c_explicit_ra              IS NOT NULL THEN c_observation_id END AS c_explicit_base_id,
+  CASE WHEN c_air_mass_min             IS NOT NULL THEN c_observation_id END AS c_air_mass_id,
+  CASE WHEN c_hour_angle_min           IS NOT NULL THEN c_observation_id END AS c_hour_angle_id,
+  CASE WHEN c_observing_mode_type      IS NOT NULL THEN c_observation_id END AS c_observing_mode_id,
+  CASE WHEN c_spec_wavelength          IS NOT NULL THEN c_observation_id END AS c_spec_wavelength_id,
+  CASE WHEN c_spec_signal_to_noise_at  IS NOT NULL THEN c_observation_id END AS c_spec_signal_to_noise_at_id,
+  CASE WHEN c_spec_wavelength_coverage IS NOT NULL THEN c_observation_id END AS c_spec_wavelength_coverage_id,
+  CASE WHEN c_spec_focal_plane_angle   IS NOT NULL THEN c_observation_id END AS c_spec_focal_plane_angle_id
+  FROM t_observation o;
+
+-- Create a view for the observation reference, including only observations with
+-- a defined reference.  This makes mapping the reference as optional easier.
+CREATE VIEW v_observation_reference AS
+  SELECT
+    o.c_observation_id,
+    o.c_program_id,
+    o.c_observation_index,
+    p.c_program_reference,
+    CONCAT(p.c_program_reference::text, '-', LPAD(o.c_observation_index::text, 4, '0')) AS c_observation_reference
```

Comment:
We'd discussed making the program reference part of a compound PK for `t_program` and then adding a FK reference in `t_observation` for which updates cascade.  It occurred to me that it might also just be computed in a view and let searching be handled by a combination of the program reference and the observation index.  What do we think?

Reply:
My thinking was that if prog/obs references are actual [computed] columns then we can do indexed searches without having to parse the input references. If we compute these values in a view then I think searches will end up being linear ... ? 

I may misunderstand your proposal here.

Reply:
Given an observation reference like "G-2025A-0001-Q-0001" you can find it, I think quickly, with

```scala
    val selectOid: Query[ObservationReference, Observation.Id] =
      sql"""
        SELECT o.c_observation_id
        FROM t_observation o
        INNER JOIN t_program p ON p.c_program_id = o.c_program_id
        WHERE p.c_program_reference = $program_reference
          AND o.c_observation_index = $int4_pos
      """.query(observation_id).contramap[ObservationReference] { ref =>
        (ref.programReference, ref.observationIndex)
      }
```

I may be overlooking the obvious.